lib/deserializer-utils.js

```diff
@@ -27,34 +24,30 @@ module.exports = function (jsonapi, data, opts) {
     }
   }
 
-  function findIncluded(relationshipData, relationshipName, from) {
+  function findIncluded(relationshipData, ancestry) {
     return new Promise(function (resolve) {
-      if (!jsonapi.included || !relationshipData) { return resolve(null); }
+      if (!jsonapi.included || !relationshipData) { resolve(null); }
 
       var included = _find(jsonapi.included, {
         id: relationshipData.id,
         type: relationshipData.type
       });
 
-      var path = [
-        from.type,
-        from.id,
-        relationshipName,
-        relationshipData.type,
-        relationshipData.id,
-      ]
-
-      // Check if the include is already processed (prevent circular
-      // references).
-      if (_get(alreadyIncluded, path, false)) {
-        return resolve(null);
-      } else {
-        _merge(alreadyIncluded, _set({}, path, true));
-      }
-
       if (included) {
+        // To prevent circular references, check if the record type
+        // has already been processed in this thread
+        if (ancestry.indexOf(included.type) > -1) {
+          return Promise
+            .all([extractAttributes(included)])
+            .then(function (results) {
+              var attributes = results[0];
+              var relationships = results[1];
+              resolve(_extend(attributes, relationships));
+            });
+        }
+
         return Promise
-          .all([extractAttributes(included), extractRelationships(included)])
+          .all([extractAttributes(included), extractRelationships(included, `${ancestry}:${included.type}${included.id}`)])
```

Comment:
It seems that Travis CI doesn't recognize ES6 template strings.  I don't know whether Travis warnings have previously been adhered to, but if it has, perhaps reverting back to the old style string concatenation or somehow installing babel might be nice?

Reply:
👍 will do